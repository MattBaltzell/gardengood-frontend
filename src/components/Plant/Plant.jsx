import React, { useState, useEffect } from "react";
import { useParams, Link } from "react-router-dom";
import GardenGoodApi from "../../api/api";
import { BallTriangle } from "react-loader-spinner";
import "./Plant.css";

const Plant = () => {
  const { id } = useParams();

  const [isLoading, setIsLoading] = useState(true);
  const [plant, setPlant] = useState({});

  useEffect(() => {
    const getPlant = async () => {
      const plant = await GardenGoodApi.getPlant(id);
      await setPlant(plant);
      setIsLoading(false);
    };
    getPlant();
  }, [id]);

  if (isLoading) {
    return (
      <main>
        <BallTriangle
          height={100}
          width={100}
          radius={5}
          color="var(--color-primary-400)"
          ariaLabel="ball-triangle-loading"
          wrapperClass={{}}
          wrapperStyle=""
          visible={true}
        />
        <p className="pulsing">Loading...</p>
      </main>
    );
  }

  return (
    <>
      <Link to="/plants" className="breadcrumb">
        Back to Plants
      </Link>
      <main className="Plant">
        <div className="Plant__info-container">
          <div className="Plant__info-header">
            <h1>{plant.name}</h1>
            <small>{plant.isPerrenial ? "Perrenial" : "Annual"}</small>
          </div>

          <div className="Plant__img-info-group">
            <img className="Plant__img" alt={plant.name} src={plant.imgUrl} />
            <div className="Plant__info-group">
              <div className="Plant__info">
                <h4>Species:</h4>
                <p>{plant.species}</p>
              </div>
              <div className="Plant__info">
                <h4>Days to Maturity</h4>
                <p>{`${plant.daysToMaturityMin} - ${plant.daysToMaturityMax}`}</p>
              </div>
              <div className="Plant__info">
                <h4>Sun Amount</h4>
                <p>{plant.sunlight.join(", ")}</p>
              </div>
              <div className="Plant__info">
                <h4>Growing Seasons</h4>
                <p>{plant.growingSeasons.join(", ")}</p>
              </div>
            </div>
          </div>

          <hr></hr>

          <div className="Plant__info">
            <h4>About</h4>
            <p>{plant.description}</p>
          </div>
          <hr></hr>
          <h2>Instructions</h2>
          <div className="Plant__instructions">
            <div className="Plant__info">
              <h4>Planting</h4>
              <p>{plant.instructions.planting || "Coming soon."}</p>
            </div>
            <div className="Plant__info">
              <h4>Pruning</h4>
              <p>{plant.instructions.pruning || "Coming soon."}</p>
            </div>
            <div className="Plant__info">
              <h4>Watering</h4>
              <p>{plant.instructions.watering || "Coming soon."}</p>
            </div>
          </div>
        </div>
      </main>
    </>
  );
};

export default Plant;
