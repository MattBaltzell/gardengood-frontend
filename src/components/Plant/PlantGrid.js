import React from "react";

const PlantGrid = ({ plants }) => {
  return (
    <>
      {plants.map((plant) => {
        return (
          <div key={plant.id} className="Plant">
            <img className="Plant-img" alt={plant.name} src={plant.imgUrl} />
            <h3>{plant.name}</h3>
          </div>
        );
      })}
    </>
  );
};

export default PlantGrid;
