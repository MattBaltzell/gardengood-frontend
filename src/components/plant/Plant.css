.Plant {
  max-width: 1200px;
  text-align: left;
  display: flex;
  width: 100%;
  flex-direction: column;
  align-items: center;
  line-height: 1.45;
}

.Plant p,
.Plant h2,
.Plant h3,
.Plant h4 {
  align-self: flex-start;
}

.Plant h2 {
  margin-bottom: -0.5rem;
}

.Plant__info-container {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: 1.5rem;
  width: 100%;
}
.Plant__img {
  object-fit: cover;
  width: 100%;
  height: 25vh;
  border-radius: 6px;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.3);
  transform: translateY(0);
  transition: all 0.2s ease-in-out;
  overflow: hidden;
}

.Plant__info-header {
  gap: 1rem;
  display: flex;
  width: 100%;
  justify-content: space-between;
  align-items: center;
  margin-bottom: -0.75rem;
}

.Plant__info-header small {
  background-color: var(--color-neutral-200);
  padding: 0.2rem 0.8rem;
  border-radius: 200px;
  font-weight: 500;
  margin-top: 4px;
  color: var(--color-neutral-600);
}

.Plant__img-info-group {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  width: 100%;
}

.Plant__info-group {
  display: grid;
  gap: 1.5rem;
  grid-template-columns: repeat(2, 1fr);
  width: 100%;
}

.Plant__info {
  width: 100%;
}

.Plant__info h4 {
  margin-bottom: 0.3rem;
  color: var(--color-primary-400);
}
.Plant__img-info-group {
  padding: 0;
}

.Plant__instructions {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

.Plant-search {
  display: flex;
  gap: 1.25rem;
  flex-direction: column;
  width: 100%;
  align-items: center;
}

.Plant-search .Plant-list {
  max-width: 1200px;
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  gap: 1rem;
  width: 100%;
}

.Plant-search__form {
  display: flex;
  flex-direction: column;
  gap: 1.2rem;
  align-items: center;
  width: 100%;
  justify-content: center;
  margin-bottom: 1rem;
  max-width: Min(90%, 400px);
}

.Plant-search__form input {
  font-size: 1rem;
  padding: 0.5rem 1rem;
  width: 100%;
  min-width: 100%;
  border: none;
  border-radius: 3px;
  color: var(--color-neutral-800);
}

.Plant-search__form input:focus {
  outline: var(--outline-primary-1);
  border-color: transparent;
}

.Plant-list__thumbnail {
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: left;
  justify-content: flex-start;
  width: 100%;
  text-align: center;
  gap: 0.5rem;
  border-radius: 3px;
  border-radius: 50%;
  margin-bottom: 0.5rem;
}

.Plant-list__name {
  font-size: 0.8rem;
  transition: all 0.2s ease-in-out;
}

.Plant-list__img {
  object-fit: cover;
  width: 100%;
  border-radius: 50%;
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.3);
  position: relative;
  transform: translateY(0);
  transition: all 0.2s ease-in-out;
}

.Plant-list__thumbnail:hover .Plant-list__img {
  filter: brightness(1.1);
  transform: translateY(-0.8rem) scale(1.03);
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.4);
  outline: var(--outline-primary-1);
}
.Plant-list__thumbnail:hover .Plant-list__name {
  color: #fff;
  transform: scale(1.1);
}

@media screen and (min-width: 530px) {
  .Plant-search .Plant-list {
    grid-template-columns: repeat(4, 1fr);
    gap: 1.5rem;
  }
}

@media screen and (min-width: 760px) {
  .Plant {
    background-color: var(--color-neutral-600);
    border-radius: 0.5rem;
    margin-top: 1rem;
    padding: 3rem;
  }

  .Plant-search .Plant-list {
    grid-template-columns: repeat(6, 1fr);
  }

  .Plant h2 {
    margin-bottom: -0.5rem;
  }

  .Plant__info-container {
    flex-direction: column;
    align-items: flex-start;
    gap: 1.5rem;
    width: 100%;
  }
  .Plant__img {
    object-fit: cover;
    width: 100%;
    height: 35vh;
    border-radius: 6px;
  }

  .Plant__img-info-group {
  }

  .Plant__info-group {
    gap: 1rem;
    padding: 1rem;
    border: 3px solid var(--color-primary-400);
    grid-template-columns: repeat(4, 1fr);
    grid-template-rows: repeat(1, 1fr);
    text-align: center;
    height: main-content;
    border-radius: 6px;
  }

  .Plant__instructions {
    flex-direction: row;
    width: 100%;
    gap: 1.5rem;
  }
}

@media screen and (min-width: 900px) {
  .Plant-search .Plant-list {
    grid-template-columns: repeat(8, 1fr);
    gap: 2rem;
  }

  .Plant__img-info-group {
    flex-direction: row-reverse;
    gap: 1rem;
  }

  .Plant__info-group {
    gap: 4rem;
    padding: 3rem;
    grid-template-rows: repeat(2, 1fr);
    grid-template-columns: repeat(2, 1fr);
    height: 35vh;
    border-radius: 6px;
  }
}
