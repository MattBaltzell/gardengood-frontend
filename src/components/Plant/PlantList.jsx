import "./Plant.css";

const PlantList = ({ plants }) => {
  return (
    <div className="Plant-list">
      {plants.map((plant) => {
        return (
          <div key={plant.id} className="Plant-list_thumbnail">
            <img
              className="Plant-list_img"
              alt={plant.name}
              src={plant.imgUrl}
            />
            <h3 className="Plant-list_name">{plant.name}</h3>
          </div>
        );
      })}
    </div>
  );
};

export default PlantList;
