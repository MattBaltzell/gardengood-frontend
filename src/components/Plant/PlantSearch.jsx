import React, { useState, useEffect } from "react";
import "./Plant.css";
import "../Form/Form.css";
import PlantList from "./PlantList";

const PlantSearch = ({ plants, search }) => {
  const INITIAL_STATE = { searchTerm: "" };
  const [formData, setFormData] = useState(INITIAL_STATE);

  const handleChange = async (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({ ...data, [name]: value }));
    search(value.trim() || undefined);
  };

  useEffect(() => {
    return () => {
      setFormData(INITIAL_STATE);
    };
  }, []);

  const handleSubmit = (e) => {
    e.preventDefault();
    search(formData.searchTerm.trim() || undefined);
  };

  return (
    <main className="Plant-search">
      <h1>Find a Plant</h1>
      <form className="Form Plant-search__form" onSubmit={handleSubmit}>
        <input
          className="Form__input--text"
          name="searchTerm"
          onChange={handleChange}
          type="text"
          value={formData.searchTerm}
          placeholder="Search plant name"
        />
      </form>
      <PlantList plants={plants} />
    </main>
  );
};

export default PlantSearch;
