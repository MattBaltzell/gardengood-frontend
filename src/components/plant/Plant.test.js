import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import Plant from "./Plant";
import UserContext from "../auth/UserContext";

const user = {
  username: "testuser",
  first_name: "firstname",
  last_name: "lastname",
  email: "test@test.com",
};

describe("Plant component tests", function () {
  it("renders without crashing", function () {
    render(
      <MemoryRouter>
        <Plant />
      </MemoryRouter>
    );
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <MemoryRouter>
        <UserContext.Provider value={user}>
          <Plant />
        </UserContext.Provider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot when logged out", function () {
    const { asFragment } = render(
      <MemoryRouter>
        <UserContext.Provider value={null}>
          <Plant />
        </UserContext.Provider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
