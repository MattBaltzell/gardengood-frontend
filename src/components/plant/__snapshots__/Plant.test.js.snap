// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plant component tests matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="NotFound"
  >
    <h1>
      Oops!
    </h1>
    <h3>
      We couldn't find that page!
    </h3>
  </main>
</DocumentFragment>
`;

exports[`Plant component tests matches snapshot when logged out 1`] = `
<DocumentFragment>
  <main
    class="NotFound"
  >
    <h1>
      Oops!
    </h1>
    <h3>
      We couldn't find that page!
    </h3>
  </main>
</DocumentFragment>
`;
